src/main/java/ch/njol/skript/lang/function/FunctionRegistry.java

```diff
@@ -0,0 +1,701 @@
+	 * stored in a {@link Retrieval} object.
+	 */
+	private @NotNull Retrieval<Function<?>> getFunction(@NotNull NamespaceIdentifier namespace, @NotNull FunctionIdentifier provided) {
+		Preconditions.checkNotNull(namespace, "namespace cannot be null");
+		Preconditions.checkNotNull(provided, "provided cannot be null");
+
+		Namespace ns = namespaces.getOrDefault(namespace, new Namespace());
+		Set<FunctionIdentifier> existing = ns.identifiers.get(provided.name);
+		if (existing == null) {
+			Skript.debug("No functions named '%s' exist in the '%s' namespace", provided.name, namespace.name);
+			return new Retrieval<>(RetrievalResult.NOT_REGISTERED, null, null);
+		}
+
+		Set<FunctionIdentifier> candidates = candidates(provided, existing);
+		if (candidates.isEmpty()) {
+			Skript.debug("Failed to find a function for '%s'", provided.name);
+			return new Retrieval<>(RetrievalResult.NOT_REGISTERED, null, null);
+		} else if (candidates.size() == 1) {
+			if (Skript.debug()) {
+				Skript.debug("Matched function for '%s': %s", provided.name, candidates.stream().findAny().orElse(null));
+			}
+			return new Retrieval<>(RetrievalResult.EXACT,
+				ns.functions.get(candidates.stream().findAny().orElse(null)),
+				null);
+		} else {
+			if (Skript.debug()) {
+				String options = candidates.stream().map(Record::toString).collect(Collectors.joining(", "));
+				Skript.debug("Failed to match an exact function for '%s'", provided.name);
+				Skript.debug("Identifier: %s", provided);
+				Skript.debug("Options: %s", options);
+			}
+			return new Retrieval<>(RetrievalResult.AMBIGUOUS,
+				null,
+				candidates.stream()
+					.map(FunctionIdentifier::args)
+					.toArray(Class<?>[][]::new));
+		}
+	}
+
+	/**
+	 * Gets the signature for a function with the given name and arguments. If no local function is found,
+	 * checks for global functions.
+	 *
+	 * <ul>
+	 * <li>If {@code namespace} is null, only global signatures will be checked.</li>
+	 * <li>If {@code args} is null or empty,
+	 * the first function with the same name as the {@code name} param will be returned.</li>
+	 * </ul>
+	 *
+	 * @param namespace The namespace to get the function from.
+	 *                  Usually represents the path of the script this function is registered in.
+	 * @param name      The name of the function.
+	 * @param args      The types of the arguments of the function.
+	 * @return The signature for the function with the given name and argument types, or null if no such function exists.
+	 */
+	public Retrieval<Signature<?>> getSignature(@Nullable String namespace, @NotNull String name, Class<?>... args) {
+		if (namespace == null) {
+			return getSignature(GLOBAL_NAMESPACE, FunctionIdentifier.of(name, false, args));
+		}
+
+		Retrieval<Signature<?>> attempt = getSignature(new NamespaceIdentifier(Scope.LOCAL, namespace),
+			FunctionIdentifier.of(name, true, args));
+		if (attempt.result != RetrievalResult.EXACT) {
```

Comment:
I would rework this to avoid calling `getSignature` unnecessarily. Something like:
```suggestion
		
		Retrieval<Signature<?>> attempt = null;
		if (namespace != null) {
			attempt = getSignature(new NamespaceIdentifier(Scope.LOCAL, namespace),
					FunctionIdentifier.of(name, true, args));
		}
		if (attempt == null || attempt.result != RetrievalResult.EXACT) {
```

Reply:
somehow i missed that